### DISCUSSION TITLE
Handling typical runtime errors

### DISCUSSION THREAD

**Discussion by mohs8421** [SeaQL/sea-orm#709] [Ideas]
## Summary
Queries can fail for various reasons, I would like to get better details from sea-orm directly instead of having to parse the error code and message from the database.

## Motivation
When an ActiveModel is saved, it can happen, that the query fails because some constraints defined in the database are not fullfilled.
For example I get something like this from a mariadb database:

> Execution Error: error returned from database: 1062 (23000): Duplicate entry 'some.duplicate@email.invalid' for key 'email'

Now I would have to parse the error code coming from the database and look up what it means in order to have my code react more specific to the error which occurred, like trying again, if the query was a deadlock-victim, or just returning the info that this already exists to the user.
The [current implementation](https://github.com/SeaQL/sea-orm/blob/master/src/error.rs) does not give me that info.
A simple implementation could be to just add an attribute to the errors which then contains the error code, so users could directly access that.
A more sophisticated variant could provide a unified interpretation of error codes of different database codes, I'm fine with any improvement, and might even think about providing my solution to the issue to the project, once I am happy with that myself.

## Additional Information
Note, that sqlx [already provides the error code separated](https://docs.rs/sqlx/latest/sqlx/error/trait.DatabaseError.html), but it is only passed as string by sea-orm

**Comment by billy1624** [comment]
Hey @mohs8421, thanks for the suggestions! Actually, this has been asked by a few members in the community. I think we can include the error code in the following three variants.
- https://github.com/SeaQL/sea-orm/blob/1a37212940205a001adcda194c600125dfc4a767/src/error.rs#L3-L9

However, I believe the error code is db dependent so user have to match the error code itself based on the db backend.

**Reply by mohs8421** [reply]
thanks, it's fine with me this way, I have already an implementation behind a feature switch at hand, it's just fully working in all given contexts yet, and it does behave a little bit different than you suggest (I added a separate Enum value for that if the feature is enabled, to avoid disrupting existing code depending on it). I can hand it over if you want it.

**Reply by billy1624** [reply]
Feel free to open an PR / post it somewhere for us :)
Thanks!!

**Reply by mohs8421** [reply]
Please have a look at this PR: https://github.com/SeaQL/sea-orm/pull/750
Convert it into which ever implementation you consider best.
I did have some warnings, but I was unsure whether I created them or they just popped up due to testing.

**Comment by jacob-pro** [comment]
For this to be useful we could really do with some improvements upstream though:

https://github.com/launchbadge/sqlx/issues/2102

There is a PR currently open:

https://github.com/launchbadge/sqlx/pull/2109

**Reply by mohs8421** [reply]
We also have a PR waiting to be continued in any way: https://github.com/SeaQL/sea-orm/pull/1110
Unfortunately I personally am currently involved in different things and don't know why this is waiting.
